### ISSUE TITLE
Shadow store support

### ISSUE THREAD

**Issue by sebastianburckhardt** [microsoft/durabletask-netherite#338]
Based on repeated user feedback, I am considering adding a feature to support **shadow stores**.

Configuring a shadow store allows users to mirror the current state of all instances (orchestrations or entities) in an external database. Shadow stores can be useful to support more types of queries (e.g. by using a database that indexes all fields or supports relational queries), or to support a high volume of instance status queries (currently a weakness of Netherite). There may be other uses as well.

The current design sketch is as follows:

- users write a piece of code that knows how to update the external store, and implements an `IShadowStore` interface, something like
```
  public interface IShadowStore
  {
      Task OnInstanceUpdated(OrchestrationInstance instanceState);
      Task OnInstanceDeleted (string instanceId);
  }
```
- this adapter is registered with the Netherite orchestration service during startup
- there are two new client operations `AddShadowStore` and `RemoveShadowStore` that allow any number of shadow stores to be registered with Netherite. 
- there is a new client operation `GetShadowStoreStatus` that can display the status of a shadow store (e.g. number of pending propagations, error messages). 

Once a shadow store is registered, Netherite automatically and persistently calls the IShadowStore methods to ensure the shadow store stays up to date (regardless of crashes, application restarts, etc). It reliably keeps track of any state updates that need to be propagated. Note that it may skip intermediate updates (e.g. if an instance state is updated multiple times it may send only the last update).

**Comment by davidmrdavid** [comment]
Nice. Is it correct to assume the shadow store can only be used for "read" operations (like getting an orchestrator status) but not for scheduling new work (i.e raising an external event)?

**Comment by sebastianburckhardt** [comment]
Right. For now, the shadow store is intended to be just a read-only eventually consistent copy.
